acme/dns_challenge.go

```diff
@@ -60,50 +61,121 @@ func (s *dnsChallenge) Solve(chlng challenge, domain string) error {
 		}
 	}()
 
-	fqdn, _, _ := DNS01Record(domain, keyAuth)
+	fqdn, value, _ := DNS01Record(domain, keyAuth)
 
-	preCheckDNS(domain, fqdn)
+	logf("[INFO][%s] Checking DNS record propagation...", domain)
+
+	if err = preCheckDNS(domain, fqdn, value); err != nil {
+		return err
+	}
 
 	return s.validate(s.jws, domain, chlng.URI, challenge{Resource: "challenge", Type: chlng.Type, Token: chlng.Token, KeyAuthorization: keyAuth})
 }
 
-func checkDNS(domain, fqdn string) bool {
-	// check if the expected DNS entry was created. If not wait for some time and try again.
-	m := new(dns.Msg)
-	m.SetQuestion(domain+".", dns.TypeSOA)
-	c := new(dns.Client)
-	in, _, err := c.Exchange(m, "google-public-dns-a.google.com:53")
+// checkDnsPropagation checks if the expected TXT record has been propagated to
+// all authoritative nameservers. If not it waits and retries for some time.
+func checkDnsPropagation(domain, fqdn, value string) error {
+	authoritativeNss, err := lookupNameservers(toFqdn(domain))
```

Comment:
> A TXT record can not co-exist with a NS record for the same name if the name is a subdomain and not the root domain. 

It can and does.

Scenario: issuing a certificate for `is-delegated.willglynn.com`.

`willglynn.com IN NS ns1.willglynn.com`, which has `willglynn.com IN SOA ns1.willglynn.com`. Within the `willglynn.com` zone, `is-delegated.willglynn.com` does not exist, but `_acme-challenge.is-delegated.willglynn.com IN NS ns1.lerfjhax.com`.

`ns1.lerfjhax.com` has a `_acme-challenge.is-delegated.willglynn.com IN SOA ns1.lerfjhax.com`, and within *that* zone, `_acme-challenge.is-delegated.willglynn.com IN NS ns1.lerfjhax.com`, agreeing with the delegation, and there exists an `_acme-challenge.is-delegated.willglynn.com IN TXT` record as well.

Asking `lookupNameservers(toFqdn(domain))` about the validation domain returns a referral, since `ns1.willglynn.com` knows it is not an authority for the validation domain:

    $ dig txt _acme-challenge.is-delegated.willglynn.com @ns1.willglynn.com

    ; <<>> DiG 9.8.3-P1 <<>> txt _acme-challenge.is-delegated.willglynn.com @ns1.willglynn.com
    ;; global options: +cmd
    ;; Got answer:
    ;; ->>HEADER<<- opcode: QUERY, status: NOERROR, id: 988
    ;; flags: qr rd; QUERY: 1, ANSWER: 0, AUTHORITY: 1, ADDITIONAL: 0
    ;; WARNING: recursion requested but not available

    ;; QUESTION SECTION:
    ;_acme-challenge.is-delegated.willglynn.com. IN TXT

    ;; AUTHORITY SECTION:
    _acme-challenge.is-delegated.willglynn.com. 300	IN NS ns1.lerfjhax.com.

    ;; Query time: 62 msec
    ;; SERVER: 205.251.198.58#53(205.251.198.58)
    ;; WHEN: Tue Feb  9 13:25:06 2016
    ;; MSG SIZE  rcvd: 87

Asking the nameserver which *is* actually an authority for the validation domain name – i.e. `lookupNameservers(toFqdn(fqdn))` – returns the TXT record:

    $ dig txt _acme-challenge.is-delegated.willglynn.com @ns1.lerfjhax.com

    ; <<>> DiG 9.8.3-P1 <<>> txt _acme-challenge.is-delegated.willglynn.com @ns1.lerfjhax.com
    ;; global options: +cmd
    ;; Got answer:
    ;; ->>HEADER<<- opcode: QUERY, status: NOERROR, id: 24113
    ;; flags: qr aa rd; QUERY: 1, ANSWER: 1, AUTHORITY: 1, ADDITIONAL: 2
    ;; WARNING: recursion requested but not available

    ;; QUESTION SECTION:
    ;_acme-challenge.is-delegated.willglynn.com. IN TXT

    ;; ANSWER SECTION:
    _acme-challenge.is-delegated.willglynn.com. 60 IN TXT "1Mvyj8Fkgch0YMz0PyTk83hrPIHrYgR-pbvN3KuevTw"

    ;; AUTHORITY SECTION:
    _acme-challenge.is-delegated.willglynn.com. 259200 IN NS ns1.lerfjhax.com.

    ;; ADDITIONAL SECTION:
    ns1.lerfjhax.com.	86400	IN	A	208.79.90.50
    ns1.lerfjhax.com.	86400	IN	AAAA	2607:f2f8:a428::2

    ;; Query time: 76 msec
    ;; SERVER: 2607:f2f8:a428::2#53(2607:f2f8:a428::2)
    ;; WHEN: Tue Feb  9 13:25:08 2016
    ;; MSG SIZE  rcvd: 187

And yes, this works:

    2016/02/09 13:25:14 [INFO][is-delegated.willglynn.com] The server validated our request
    2016/02/09 13:25:14 [INFO] acme: You can now remove this TXT record from your DNS zone:
    2016/02/09 13:25:14 [INFO] acme: _acme-challenge.is-delegated.willglynn.com. 120 IN TXT "..."
    2016/02/09 13:25:14 [INFO][is-delegated.willglynn.com] acme: Validations succeeded; requesting certificates
    2016/02/09 13:25:14 [INFO] acme: Requesting issuer cert from https://acme-staging.api.letsencrypt.org/acme/issuer-cert
    2016/02/09 13:25:14 [INFO][is-delegated.willglynn.com] Server responded with a certificate.

> But again, this discussion is out of scope for this PR as Lego's DNS providers don't delegate the validation domain anywhere.

…unless you delegate the validation domain ahead of time only need `lego` to create the TXT record in the delegated zone. As I said, `RFC2136_ZONE=_acme-challenge.foo.bar.com lego --dns=rfc2136 ` can do that right now.